### PR TITLE
Update Java Spark bindings for CB support

### PR DESCRIPTION
**Pull request by jackgerrits**

### CODE REVIEW

File: java/src/main/c++/jni_spark_vw.cc [hunk 1]
```diff
@@ -77,6 +77,85 @@ JNIEXPORT jlong JNICALL Java_org_vowpalwabbit_spark_VowpalWabbitNative_initializ
...
+
+      ex_coll.push_back(exWrapper->_example);
+    }
+  }
+}
+
+JNIEXPORT jobject JNICALL Java_org_vowpalwabbit_spark_VowpalWabbitNative_learn(
+    JNIEnv* env, jobject vwObj, jobjectArray examples)
+{
+  auto all = (vw*)get_native_pointer(env, vwObj);
```
Comment (olgavrou) [inline java/src/main/c++/jni_spark_vw.cc:113]: static_cast?

Comment (jackgerrits) [reply java/src/main/c++/jni_spark_vw.cc:113]: It'll need to be reinterpret_cast, but you're right shouldn't be a cstyle cast

File: java/src/main/c++/jni_spark_vw_generated.h [hunk 2]
```diff
@@ -23,6 +23,20 @@ extern "C"
   JNIEXPORT jlong JNICALL Java_org_vowpalwabbit_spark_VowpalWabbitNative_initializeFromModel(
```
[the remark this answers is not in the document]
Comment (jackgerrits) [reply java/src/main/c++/jni_spark_vw_generated.h:23]: @eisber thoughts?

File: java/src/main/c++/jni_spark_vw.cc (hunk 1 above)
Comment (olgavrou) [reply java/src/main/c++/jni_spark_vw.cc:113]: why won't static cast do?

Comment (olgavrou) [reply java/src/main/c++/jni_spark_vw.cc:113]: ah the function returns an integer, ok

File: java/src/main/c++/jni_spark_vw_generated.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (eisber) [reply java/src/main/c++/jni_spark_vw_generated.h:23]: I vaguely remember that there was some cyclic dependency what needs to be there first (in terms of build), but maybe it was just me...

File: java/src/main/c++/jni_spark_vw.cc [hunk 3]
```diff
@@ -123,12 +202,14 @@ JNIEXPORT jbyteArray JNICALL Java_org_vowpalwabbit_spark_VowpalWabbitNative_getM
   catch (...)
   {
     rethrow_cpp_exception_as_java_exception(env);
+    // return null
+    return 0;
```
Comment (lalo) [inline java/src/main/c++/jni_spark_vw.cc:206]: nullptr?

File: java/src/main/c++/jni_spark_vw.cc [hunk 4]
```diff
@@ -415,62 +498,70 @@ JNIEXPORT void JNICALL Java_org_vowpalwabbit_spark_VowpalWabbitExample_setLabel(
...
+  try
+  {
+    // https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/vowpalwabbit/parse_example_json.h#L437
+    CB::label* ld = &ex->l.cb;
+    CB::cb_class f;
 
-    case prediction_type_t::action_probs:
-      return action_probs_prediction(ex, env);
+    f.partial_prediction = 0.;
+    f.action = (uint32_t)uniform_hash("shared", 6, 0);
```
Comment (lalo) [inline java/src/main/c++/jni_spark_vw.cc:548]: maybe add comment referencing that 6 is length of shared and 0 is the seed. This is an ugly api but not part of this PR.

Comment (lalo) [reply java/src/main/c++/jni_spark_vw.cc:548]: totally optional though, not sure how many people run into this code or whats the probability of someone changing shared to another string

File: java/src/main/c++/jni_spark_vw.cc [hunk 5]
```diff
@@ -505,10 +596,145 @@ JNIEXPORT jobject JNICALL Java_org_vowpalwabbit_spark_VowpalWabbitExample_predic
...
+        if ((ns >= 'a' && ns <= 'z') || (ns >= 'A' && ns <= 'Z'))
+          ostr << "'" << (char)ns << "':";
+
+        ostr << (int)ns << ",";
+      }
+
+      for (auto& f : ex->feature_space[ns])
+      {
+        auto idx = f.index();
+        ostr << (idx & all->weights.mask()) << "/" << idx << ":" << f.value() << ", ";
```
Comment (lalo) [inline java/src/main/c++/jni_spark_vw.cc:664]: I'm guessing we're resilient to ending it with an extra comma?

File: java/src/main/c++/jni_spark_vw.cc [hunk 6]
```diff
@@ -505,10 +596,145 @@ JNIEXPORT jobject JNICALL Java_org_vowpalwabbit_spark_VowpalWabbitExample_predic
...
+      CB::label* ld = (CB::label*)&ex->l;
+      ostr << "CB " << ld->costs.size();
+
+      if (ld->costs.size() > 0)
+      {
+        ostr << " ";
+
+        CB::cb_class& f = ld->costs[0];
+
+        // ignore validation of action which is uniform_hash("shared")
```
Comment (lalo) [inline java/src/main/c++/jni_spark_vw.cc:636]: what does this comment mean? where are we ignoring it exactly

File: java/src/main/c++/jni_spark_vw.cc (hunk 3 above)
Comment (jackgerrits) [reply java/src/main/c++/jni_spark_vw.cc:206]: Will do.

Went and checked and jbytearray is defined as:
```C++
typedef _jbyteArray *jbyteArray;
```

File: java/src/main/c++/jni_spark_vw.cc (hunk 4 above)
Comment (jackgerrits) [reply java/src/main/c++/jni_spark_vw.cc:548]: Not sure why the action must be this value... But this was copied from elsewhere

File: java/src/main/c++/jni_spark_vw.cc (hunk 5 above)
Comment (jackgerrits) [reply java/src/main/c++/jni_spark_vw.cc:664]: its just a debug string so its okay

File: java/src/main/c++/jni_spark_vw.cc (hunk 3 above)
Comment (lalo) [reply java/src/main/c++/jni_spark_vw.cc:206]: theres some others, but not all of them. i saw a long in there or something but it also might be a pointer

File: java/src/main/c++/jni_spark_vw.cc (hunk 6 above)
Comment (jackgerrits) [reply java/src/main/c++/jni_spark_vw.cc:636]: Updated comment. Means we don't check f.action
